Tools/CISupport/ews-build/steps.py

```diff
@@ -2150,6 +2174,188 @@ def hideStepIf(self, results, step):
+                        if c['state'] == "FAILURE" and c['context'] in check_tests:
+                            failed_status_check.append(pr_num)
+                            failed = True
+                            yield self._addToLog('stdio', 'Failed status check. Done!\n')
+                            break
+                    if not failed:
+                        passed_status_check.append(pr_num)
+                        yield self._addToLog('stdio', 'Passed status check. Done!\n')
+
+        defer.returnValue([list_of_prs, passed_status_check, failed_status_check])
+
+
+class SafeMergeToMergeQueue(buildstep.BuildStep, GitHubMixin, AddToLogMixin):
+    name = 'safe-merge-to-merge-queue'
+    flunkOnFailure = False
+    haltOnFailure = False
+    LABELS_TO_REMOVE = [GitHubMixin.SAFE_MERGE_QUEUE_LABEL]
+
+    @defer.inlineCallbacks
+    def run(self):
+        safe_prs = self.getProperty('safe_pr_list')
+        pr_number = safe_prs.pop()
+        yield self._addToLog('stdio', 'Updating labels for PR {}...\n'.format(pr_number))
+        self.setProperty('safe_pr_list', safe_prs)
+        if len(safe_prs):
+            self.build.addStepsAfterCurrentStep([SafeMergeToMergeQueue()])
+        rc = yield self.update_labels(pr_number)
+        defer.returnValue(rc)
+
+    def getResultSummary(self):
+        if self.results == SUCCESS:
+            return {'step': f"Labelled safe PRs with merge-queue."}
+        elif self.results == FAILURE:
+            return {'step': f"Failed to label safe PRs with merge-queue."}
+        return buildstep.BuildStep.getResultSummary(self)
+
+    @defer.inlineCallbacks
+    def update_labels(self, pr_number):
+        repository_url = self.getProperty('repository', '')
+
+        did_remove_labels = yield self.remove_labels(pr_number, self.LABELS_TO_REMOVE, repository_url=repository_url)
+        if did_remove_labels:
+            yield self._addToLog('stdio', 'Successfully removed existing labels from PR {}.\n'.format(pr_number))
+        else:
+            yield self._addToLog('stdio', 'Failed to remove existing labels from PR {}.\n'.format(pr_number))
+
+        did_add_label = yield self.add_label(pr_number, self.SAFE_MERGE_QUEUE_LABEL, repository_url=repository_url)
+        if did_add_label:
+            yield self._addToLog('stdio', 'Successfully added merge-queue label to PR {}.\n'.format(pr_number))
+        else:
+            yield self._addToLog('stdio', 'Failed to add merge-queue label to PR {}.\n'.format(pr_number))
+        defer.returnValue(SUCCESS if did_remove_labels and did_add_label else FAILURE)
+
+
+class SafeMergeToMergeBlocked(buildstep.BuildStep, GitHubMixin, AddToLogMixin):
+    name = 'safe-merge-to-merge-blocked'
+    flunkOnFailure = False
+    haltOnFailure = False
+    LABELS_TO_REMOVE = [GitHubMixin.SAFE_MERGE_QUEUE_LABEL]
+
+    @defer.inlineCallbacks
+    def run(self):
+        failed_prs = self.getProperty('blocked_pr_list', [])
+        pr_number = failed_prs.pop()
```

Comment:
Theoretically, this method would never run if `failed_status_check` is empty. But would it be better to include the check in case of some error?

Reply:
It's still better to do the error check here, we dont' want to blow up (fail the build with an exception) in case that property is not set for some unexpected reason, or someone modifies the other step without realizing the dependency here.